wasm_src/gsl_wrapper/gsl_wrapper.cc

```diff
@@ -233,4 +240,363 @@ int EMSCRIPTEN_KEEPALIVE filterSavitzkyGolay(double* xInArray, double* yInArray,
     return status;
 }
 
+struct data
+{
+  double *t;
+  double *y;
+  size_t n;
+  size_t component;
+  int function;
+  double **inputs;
+  double *orderInputs;
+  int **lockedInputs;
+  int *lockedOrderInputs;
+  gsl_matrix *parameterIndexs;
+  gsl_vector *orderParameterIndexes;
+};
+
+char logBuffer[4096];
+
+/* model function: amp * exp( -4ln2[(t - center)/ fwhm]^2 */
+double gaussian(const double amp, const double center, const double fwhm, const double x)
+{
+  return (amp * exp(-4 * log(2.0) * pow((x - center) / fwhm, 2)));
+}
+
+/* model function: amp * (0.5fwhm)^2 / [(t - center)^2 + (0.5fwhm)^2] */
+double lorentzian(const double amp, const double center, const double fwhm, const double x)
+{
+  return (amp * 0.25 * pow(fwhm, 2) / (pow(x - center, 2) + 0.25 * pow(fwhm, 2)));
+}
+
+int func_f (const gsl_vector * x, void *params, gsl_vector * f) {
+    struct data *d = (struct data *) params;
+    size_t i, j;
+    for (i = 0; i < d->n; ++i)
+    {
+        double ti = d->t[i];
+        double yi = d->y[i];
+        double y = 0;
+        for (j = 0; j <  d->component; ++j )
+        {
+            int *lockedInput = d->lockedInputs[j];
+            double amp, center, fwhm;
+            if (lockedInput[0] == 0) {
+                amp = gsl_vector_get(x, gsl_matrix_get(d->parameterIndexs, j, 0));
+            } else {
+                amp = d->inputs[j][0];
+            }
+
+            if (lockedInput[1] == 0) {
+                center = gsl_vector_get(x, gsl_matrix_get(d->parameterIndexs, j, 1));
+            } else {
+                center = d->inputs[j][1];
+            }
+
+            if (lockedInput[2] == 0) {
+                fwhm = gsl_vector_get(x, gsl_matrix_get(d->parameterIndexs, j, 2));
+            } else {
+                fwhm = d->inputs[j][2];
+            }
+
+            if (d->function == 0) {
+                y = y + gaussian(amp, center, fwhm, ti);
+            } else if (d->function == 1) {
+                y = y + lorentzian(amp, center, fwhm, ti);
+            }
+        }
+
+        double yIntercept, slope;
+        if (d->lockedOrderInputs[0] == 0) {
+            yIntercept = gsl_vector_get(x, gsl_vector_get(d->orderParameterIndexes, 0));
+        } else {
+            yIntercept = d->orderInputs[0];
+        }
+        if (d->lockedOrderInputs[1] == 0) {
+            slope = gsl_vector_get(x, gsl_vector_get(d->orderParameterIndexes, 1));
+        } else {
+            slope = d->orderInputs[1];
+        }
+
+        y = y + slope * ti + yIntercept;
+        gsl_vector_set(f, i, yi - y);
+    }
+
+    return GSL_SUCCESS;
+}
+
+void
+callback(const size_t iter, void *params,
+         const gsl_multifit_nlinear_workspace *w)
+{
+  gsl_vector *f = gsl_multifit_nlinear_residual(w);
+  gsl_vector *x = gsl_multifit_nlinear_position(w);
+  double avratio = gsl_multifit_nlinear_avratio(w);
+  double rcond;
+  size_t i;
+
+  (void) params; /* not used */
```

Comment:
what is this line for?

Reply:
deleted. This function is the legacy code from the fitting example of GSL.